File: hummingbird-client/src/main/java/com/vaadin/client/bootstrap/Bootstrapper.java
/*
 * Copyright 2000-2016 Vaadin Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */
package com.vaadin.client.bootstrap;

import com.google.gwt.core.client.EntryPoint;
import com.google.gwt.core.client.GWT;
import com.google.gwt.core.client.Scheduler;
import com.vaadin.client.ApplicationConfiguration;
import com.vaadin.client.ApplicationConnection;
import com.vaadin.client.Profiler;
import com.vaadin.client.ValueMap;
import com.vaadin.client.WidgetUtil;
import com.vaadin.client.hummingbird.collection.JsArray;
import com.vaadin.client.hummingbird.collection.JsCollections;
import com.vaadin.shared.ApplicationConstants;

import elemental.client.Browser;

/**
 * Handles bootstrapping of the application.
 * <p>
 * Reads the configuration provided by the server in the DOM and starts the
 * client engine ({@link ApplicationConnection}).
 * <p>
 * Acts as the GWT entry point.
 *
 * @author Vaadin
 * @since
 */
public class Bootstrapper implements EntryPoint {

    private static boolean moduleLoaded = false;

    private static JsArray<ApplicationConnection> runningApplications = JsCollections
            .array();

    @Override
    public void onModuleLoad() {
        initModule();
    }

    private static void initModule() {
        // Don't run twice if the module has been inherited several times,
        // and don't continue if vaadinBootstrap was not executed.
        if (moduleLoaded || !vaadinBootstrapLoaded()) {
            Browser.getWindow().getConsole().warn(
                    "vaadinBootstrap.js was not loaded, skipping vaadin application configuration.");
            return;
        }
        moduleLoaded = true;

        Profiler.initialize();

        registerCallback(GWT.getModuleName());
    }

    /**
     * Starts the application with a given id by reading the configuration
     * options stored by the bootstrap javascript.
     *
     * @param applicationId
     *            id of the application to load, this is also the id of the html
     *            element into which the application should be rendered.
     */
    public static void startApplication(final String applicationId) {
        Scheduler.get().scheduleDeferred(() -> {
            Profiler.enter("Bootstrapper.startApplication");
            ApplicationConfiguration appConf = getConfigFromDOM(applicationId);
            ApplicationConnection applicationConnection = new ApplicationConnection(
                    appConf);
            runningApplications.push(applicationConnection);
            Profiler.leave("Bootstrapper.startApplication");

            ValueMap initialUidl = getJsoConfiguration(applicationId).getUIDL();
            applicationConnection.start(initialUidl);
        });
    }

    /**
     * Gets a list of references to all running application instances.
     *
     * @return a list of ApplicationConnections currently active
     */
    public static JsArray<ApplicationConnection> getRunningApplications() {
        return runningApplications;
    }

    /**
     * Constructs an ApplicationConfiguration object based on the information
     * available in the DOM.
     *
     * @param appId
     *            the application id
     * @return an application configuration object containing the read
     *         information
     */
    private static ApplicationConfiguration getConfigFromDOM(String appId) {
        ApplicationConfiguration conf = new ApplicationConfiguration();
        conf.setApplicationId(appId);
        populateApplicationConfiguration(conf, getJsoConfiguration(appId));
        return conf;
    }

    /**
     * Reads the configuration values defined by the bootstrap Javascript.
     *
     * @param conf
     */
    private static void populateApplicationConfiguration(
            ApplicationConfiguration conf, JsoConfiguration jsoConfiguration) {
        String serviceUrl = jsoConfiguration
                .getConfigString(ApplicationConstants.SERVICE_URL);

        if (serviceUrl == null || "".equals(serviceUrl)) {
            /*
             * Use the current url without query parameters and fragment as the
             * default value.
             */
            serviceUrl = Browser.getWindow().getLocation().getHref()
                    .replaceFirst("[?#].*", "");
        } else {
            /*
             * Resolve potentially relative URLs to ensure they point to the
             * desired locations even if the base URL of the page changes later
             * (e.g. with pushState)
             */
            serviceUrl = WidgetUtil.getAbsoluteUrl(serviceUrl);
        }

        // Ensure there's an ending slash (to make appending e.g. UIDL work)
        if (!serviceUrl.endsWith("/")) {
            serviceUrl += Character.toString('/');
        }
        conf.setServiceUrl(serviceUrl);

        conf.setVaadinDirUrl(WidgetUtil.getAbsoluteUrl(jsoConfiguration
                .getConfigString(ApplicationConstants.VAADIN_DIR_URL)));
        conf.setUIId(jsoConfiguration
                .getConfigInteger(ApplicationConstants.UI_ID_PARAMETER)
                .intValue());

        // null -> false
        conf.setStandalone(jsoConfiguration.getConfigBoolean("standalone"));

        conf.setHeartbeatInterval(
                jsoConfiguration.getConfigInteger("heartbeatInterval"));

        conf.setServletVersion(jsoConfiguration.getVaadinVersion());
        conf.setAtmosphereVersion(jsoConfiguration.getAtmosphereVersion());
        conf.setAtmosphereJSVersion(jsoConfiguration.getAtmosphereJSVersion());
        conf.setCommunicationError(
                jsoConfiguration.getConfigError("comErrMsg"));
        conf.setAuthorizationError(
                jsoConfiguration.getConfigError("authErrMsg"));
        conf.setSessionExpiredError(
                jsoConfiguration.getConfigError("sessExpMsg"));

        // Debug or production mode?
        conf.setDebugMode(isDebugMode());
    }

    /**
     * Checks if debug mode is active (not disabled from server and debug flag
     * present).
     *
     * @return {@code true} if debug mode is active, {@code false} otherwise
     */
    private static boolean isDebugMode() {
        return isDebugAvailable() && getParameter("debug") != null;
    }

    private static String getParameter(String parameter) {
        String[] keyValues = Browser.getDocument().getLocation().getSearch()
                .substring(1).split("&");
        for (String keyValue : keyValues) {
            String[] param = keyValue.split("=", 1);
            if (param[0].equals(parameter)) {
                return param[1];
            }
        }

        return null;
    }

    private static native boolean isDebugAvailable()
    /*-{
        if($wnd.vaadin.debug) {
            return true;
        } else {
            return false;
        }
    }-*/;

    /**
     * Gets the configuration object for a specific application from the
     * bootstrap javascript.
     *
     * @param appId
     *            the id of the application to get configuration data for
     * @return a native javascript object containing the configuration data
     */
    private static native JsoConfiguration getJsoConfiguration(String appId)
    /*-{
        return $wnd.vaadin.getApp(appId);
     }-*/;

    private static native boolean vaadinBootstrapLoaded()
    /*-{
         return $wnd.vaadin != null;
     }-*/;

    /**
     * Registers the callback that the bootstrap javascript uses to start
     * applications once the widgetset is loaded and all required information is
     * available.
     *
     * @param widgetsetName
     *            the name of this widgetset
     */
    public static native void registerCallback(String widgetsetName)
    /*-{
        var callbackHandler = $entry(@com.vaadin.client.bootstrap.Bootstrapper::startApplication(Ljava/lang/String;));
        $wnd.vaadin.registerWidgetset(widgetsetName, callbackHandler);
    }-*/;

}

File: flow-client/src/main/frontend/internal/Bootstrapper.ts
/*
 * Copyright 2000-2026 Vaadin Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */

// Bootstrap-sequence helpers migrated from Bootstrapper.java, registered on
// window.Vaadin.Flow.internal.Bootstrapper by registerInternals; the Java
// methods delegate here. The callbacks into the GWT entry point
// (doStartApplication / startApplication) are supplied from the Java side
// already wrapped in $entry. vaadinBootstrapLoaded stays in Java: it checks
// $wnd.Vaadin.Flow, which registerInternals itself creates, so a delegating
// version would always report true. Also bundled to ES5 for the HtmlUnit used
// by GwtTests.
//
// populateApplicationConfiguration below is the build-alongside TS port of the
// Bootstrapper DOM-config reader: it fills an ApplicationConfiguration from the
// bootstrap JsoConfiguration. doStartApplication is the cutover entry: it reads
// the DOM config, assembles the TS engine via ApplicationConnection.create, and
// starts it from the initial UIDL. The Java Bootstrapper.doStartApplication
// delegates here, so the running engine is the TypeScript one.

import { ApplicationConfiguration } from './ApplicationConfiguration';
import {
  getAtmosphereVersion,
  getConfigBoolean,
  getConfigError,
  getConfigString,
  getConfigStringArray,
  getConfigValueMap,
  getVaadinVersion
} from './JsoConfiguration';
import { getAbsoluteUrl } from './WidgetUtil';

// com.vaadin.flow.shared.ApplicationConstants
const SERVICE_URL = 'serviceUrl';
const APP_WC_MODE = 'webComponentMode';
const CONTEXT_ROOT_URL = 'contextRootUrl';
const UI_ID_PARAMETER = 'v-uiId';
const DEV_TOOLS_ENABLED = 'devToolsEnabled';

/** The bootstrap configuration object's typed accessors (a JS overlay in Java). */
export interface JsoConfiguration {
  getConfigString(name: string): string | null;
  getConfigBoolean(name: string): boolean;
  getConfigInteger(name: string): number;
  getConfigStringArray(name: string): string[];
  getConfigError(name: string): unknown;
  getVaadinVersion(): string;
  getAtmosphereVersion(): string;
  getAtmosphereJSVersion(): string;
}

/**
 * Fills the application configuration from the bootstrap JavaScript config.
 * Mirrors Bootstrapper.populateApplicationConfiguration.
 */
export function populateApplicationConfiguration(
  conf: ApplicationConfiguration,
  jsoConfiguration: JsoConfiguration
): void {
  // Resolve potentially relative URLs now so they survive later base-URL changes.
  const serviceUrl = jsoConfiguration.getConfigString(SERVICE_URL);

  conf.setWebComponentMode(jsoConfiguration.getConfigBoolean(APP_WC_MODE));

  if (serviceUrl === null) {
    conf.setServiceUrl(getAbsoluteUrl('.'));
    conf.setContextRootUrl(getAbsoluteUrl(jsoConfiguration.getConfigString(CONTEXT_ROOT_URL) ?? ''));
  } else {
    conf.setServiceUrl(serviceUrl);
    conf.setContextRootUrl(getAbsoluteUrl(serviceUrl + (jsoConfiguration.getConfigString(CONTEXT_ROOT_URL) ?? '')));
  }

  conf.setUIId(jsoConfiguration.getConfigInteger(UI_ID_PARAMETER));
  conf.setHeartbeatInterval(jsoConfiguration.getConfigInteger('heartbeatInterval'));
  conf.setMaxMessageSuspendTimeout(jsoConfiguration.getConfigInteger('maxMessageSuspendTimeout'));

  conf.setServletVersion(jsoConfiguration.getVaadinVersion());
  conf.setAtmosphereVersion(jsoConfiguration.getAtmosphereVersion());
  conf.setAtmosphereJSVersion(jsoConfiguration.getAtmosphereJSVersion());
  conf.setSessionExpiredError(jsoConfiguration.getConfigError('sessExpMsg'));

  // Debug or production mode?
  conf.setProductionMode(!jsoConfiguration.getConfigBoolean('debug'));
  conf.setRequestTiming(jsoConfiguration.getConfigBoolean('requestTiming'));
  conf.setExportedWebComponents(jsoConfiguration.getConfigStringArray('webcomponents'));

  conf.setDevToolsEnabled(jsoConfiguration.getConfigBoolean(DEV_TOOLS_ENABLED));
  conf.setLiveReloadUrl(jsoConfiguration.getConfigString('liveReloadUrl') ?? '');
  conf.setLiveReloadBackend(jsoConfiguration.getConfigString('liveReloadBackend') ?? '');
  conf.setSpringBootLiveReloadPort(jsoConfiguration.getConfigString('springBootLiveReloadPort') ?? '');
}

// The raw bootstrap config object (exposes getConfig(name)); see JsoConfiguration.
interface RawConfigObject {
  getConfig(name: string): unknown;
}

/** The loaded Atmosphere library version, or '' if push is not loaded. */
function atmosphereJsVersion(): string {
  const push = (window as unknown as { vaadinPush?: { atmosphere?: { version?: string } } }).vaadinPush;
  return push?.atmosphere?.version ?? '';
}

/**
 * Wraps the raw bootstrap config object as a JsoConfiguration. getConfigInteger
 * and getAtmosphereJSVersion are inlined here (they stay Java-native in the GWT
 * build for deferred-binding reasons; see JsoConfiguration.ts).
 */
function toJsoConfiguration(rawConfig: unknown): JsoConfiguration {
  const config = rawConfig as RawConfigObject;
  return {
    getConfigString: (name) => getConfigString(config, name),
    getConfigBoolean: (name) => getConfigBoolean(config, name),
    getConfigInteger: (name) => {
      const value = config.getConfig(name);
      return value === null || value === undefined ? 0 : Number(value);
    },
    getConfigStringArray: (name) => (getConfigStringArray(config, name) as string[] | null) ?? [],
    getConfigError: (name) => getConfigError(config, name),
    getVaadinVersion: () => getVaadinVersion(config) ?? '',
    getAtmosphereVersion: () => getAtmosphereVersion(config) ?? '',
    getAtmosphereJSVersion: () => atmosphereJsVersion()
  };
}

/**
 * Starts the application with the given id: reads its configuration from the DOM,
 * assembles the TypeScript engine, and starts it from the initial UIDL. Mirrors
 * Bootstrapper.doStartApplication (which delegates here).
 */
export function doStartApplication(applicationId: string): void {
  const rawConfig = getJsoConfiguration(applicationId);

  const conf = new ApplicationConfiguration();
  conf.setApplicationId(applicationId);
  populateApplicationConfiguration(conf, toJsoConfiguration(rawConfig));

  const initialUidl = getConfigValueMap(rawConfig as RawConfigObject, 'uidl') as Record<string, unknown> | null;

  // Load the engine lazily: this keeps ApplicationConnection/DefaultRegistry and
  // the rest of the modern-JS engine out of the registerInternals bundle, which
  // the HtmlUnit-based GwtTests also load and cannot run (no Array.from, etc.).
  // The engine is only needed once a real application starts.
  void import('./ApplicationConnection').then(({ ApplicationConnection }) => {
    ApplicationConnection.create(conf).start(initialUidl ?? null);
  });
}

interface WebComponentsGlobal {
  WebComponents?: { ready?: boolean };
}

interface FlowWidgetsetRegistrar {
  Vaadin: { Flow: { registerWidgetset: (widgetsetName: string, callback: (applicationId: string) => void) => void } };
}

interface FlowAppLookup {
  Vaadin: { Flow: { getApp: (appId: string) => unknown } };
}

/**
 * Whether the application can be started immediately, i.e. there is no
 * WebComponents polyfill still loading.
 */
export function startApplicationImmediately(): boolean {
  const webComponents = (window as unknown as WebComponentsGlobal).WebComponents;
  return !webComponents || webComponents.ready === true;
}

/**
 * Defers starting the application until the WebComponents polyfill signals it
 * is ready, by running the (already $entry-wrapped) callback on the
 * WebComponentsReady event.
 */
export function deferStartApplication(callback: () => void): void {
  window.addEventListener('WebComponentsReady', callback);
}

/**
 * Registers the widgetset start callback with the bootstrap JavaScript so it can
 * start applications once the widgetset is loaded.
 */
export function registerCallback(widgetsetName: string, callback: (applicationId: string) => void): void {
  (window as unknown as FlowWidgetsetRegistrar).Vaadin.Flow.registerWidgetset(widgetsetName, callback);
}

/**
 * The bootstrap configuration object for the application with the given id, as
 * stored by the bootstrap JavaScript.
 */
export function getJsoConfiguration(appId: string): unknown {
  return (window as unknown as FlowAppLookup).Vaadin.Flow.getApp(appId);
}
